### DISCUSSION TITLE
Clarification on documentation: does `stumpy.match` rely on `MPDist`?

### DISCUSSION THREAD

**Discussion by patrickfleith** [TDAmeritrade/stumpy#799] [General]
I would like to understand better how the methods `stumpy.match` and `stumpy.mass` compute the distance under the hood to find the best matches. I assume it relies on the measure `MPDist` but it is not explicitly mention in the documentation which distance measure it uses. Could you clarify?

**Comment by seanlaw** [comment]
Hi @patrickfleith! Welcome to the STUMPY community and thank you for your question.

> I would like to understand better how the methods stumpy.match and stumpy.mass compute the distance under the hood to find the best matches. I assume it relies on the measure MPDist but it is not explicitly mention in the documentation which distance measure it uses. Could you clarify?

No, `stumpy.match` uses `stumpy.mass` to compute the distance under the hood. For all intents and purposes, `stumpy.mass` is just a fast way (leveraging FFT) to compute the sliding window dot product between the query subsequence, `Q`, and all subsequences within your time series, `T`. Once the dot products are computed, it is trivial to compute the correspodning (z-normalized) Euclidean distance(s). I recommend looking at [this paper](https://www.cs.ucr.edu/~eamonn/PID4481997_extend_Matrix%20Profile_I.pdf) to better understand the inner workings. At the end of the day, it computes the (z-normalized) Euclidean distance and that is all you should need to care about.

To be explicit, `stumpy.match` and `stumpy.mass` do not use `MPdist` and is mostly unrelated. I hope that helps.

**Comment by patrickfleith** [comment]
Great, many thanks. And from what I observe there is no other routine which search for the best matching based on MPDsit, right? One would have to come up with his.her own search function?

**Comment by seanlaw** [comment]
> Great, many thanks. And from what I observe there is no other routine which search for the best matching based on MPDsit, right? One would have to come up with his.her own search function?

That is correct. At the end of the day, you are computing a "distance" and then it is up to you to set a distance threshold that decides whether something is "significant" or not.

**Comment by patrickfleith** [comment]
Ok great thanks I close the issue

**Comment by seanlaw** [comment]
@patrickfleith In the future, would you mind posting usage questions to our [Github Discussions](https://github.com/TDAmeritrade/stumpy/discussions)? This way, others may chime in and provide their perspective too.

**Comment by NimaSarajpoor** [comment]
@patrickfleith 
I am late to the party, but please allow me to share my opinions with you. As @seanlaw mentioned, the distance measure/metric and the threshold is up to you. At the end of the day, we need that to find the best matches.

I think `stumpy` is designed in a way that you can usually find a way to (easily?) build what you are looking for. Here is what I think you can do to get what you want. (I haven't tried it but you can play with it and you may get what you are looking for)


Let's start with function `match` in `stumpy/motifs.py`:
https://github.com/TDAmeritrade/stumpy/blob/92ef4d56f4cf875638077cf595ebd8abfd4b7499/stumpy/motifs.py#L354-L367

So, this takes `Q` and `T` as the input, and then find the distance between `Q` and all subsequences of `T`, using `mass` algorithm. If you check out the function `match`, you can see this block of lines: 

https://github.com/TDAmeritrade/stumpy/blob/92ef4d56f4cf875638077cf595ebd8abfd4b7499/stumpy/motifs.py#L488-L493

As shown above, in line 490, function `mass` is used. As a side note, notice that the index `i` is NOT the index in a time series, but instead, it is the index of dimension when you have multi-dimensional data. In most cases, data is 1D, and so, `d==1 `, and hence `for i in range(d)` is basically one iteration. 

Okay...let's get back to the main point. Now, instead of `mass`, you should be able to use any function that takes input `Q` and `T` and return their distance profile. (for more info on this, take a look at the function `mass` and its doctoring). Now, all you want is to replace this with a function that is based on `mpdist` (I think that should work. right?).

Fortunately, `stumpy` supports this part of puzzle as well. If you take a look at function `_mpdist_vec` in `stumpy/mpdist.py`, you can see the following:

https://github.com/TDAmeritrade/stumpy/blob/92ef4d56f4cf875638077cf595ebd8abfd4b7499/stumpy/mpdist.py#L166-L179

Based on my understanding, this function calculates the distance between Q and all subseqs of `T` (similar to `mass`) and the distance between `Q` and each subsequence is actually MPDIST.

**Comment by patrickfleith** [comment]
@NimaSarajpoor Thanks for your great feedback. Indeed this sounds like a good way to implement what I was looking for. Thank you, I'll post the outcomes when I get back to it.

And just to give a bit more context to readers of this thread, the reason why I was asking this: it is because the `MPDist` measure has some properties that the euclidean distance doesn't (some extra invariances).
